File: test/testSGD.lua
require 'models.RNN'
require 'optimization.sgd'

rnn={i2h=nil,h2h=nil,h2o=nil,s=nil,ds=nil,buffer=0,
__init__=RNN.__init__,run1Token=RNN.run1Token,runTokens=RNN.runTokens,
updateUWV=RNN.updateUWV,updateS=RNN.updateS}
N=5
H=5
K=5
S=50
iters=1000

rnn:__init__(N,H,K)
-- print('input',rnn.i2h.weight)
-- print('recur',rnn.h2h.weight)
-- print('output',rnn.h2o.weight)

initial=torch.randn(H)
states=torch.randn(S,N)
ground_truth=torch.zeros(S,K)
for i=1,S do
	spot=math.floor(math.random()*K)+1
	ground_truth[i][spot]=1
end

time=0
for i=1,iters do
	local begin=os.clock()
	local err=sgd(rnn,states,ground_truth)
	if i%2==0 then
		rnn:updateUWV(0.1)
		rnn:updateS(0.1)
	end
	local finish=os.clock()
	time=time+finish-begin
	print(i,err)
end
print('time',time)

File: Python/test/testSGD.py
import sys
sys.path.insert(0,'util/')
sys.path.insert(0,'models/')
sys.path.insert(0,'methods/')

import RNN
import sgd_const_lr as sgd
import random
import time
import numpy as np

N=10
H=5
K=10
S=50
iters=1000

rnn=RNN.RNN(N,H,K)

states=np.random.random([S,N])*2-1
ground_truth=np.zeros([S,K])

for i in xrange(S):
    spot=random.randint(0,K-1)
    ground_truth[i,spot]=1

duration=0
for i in xrange(iters):
    begin=time.time()
    err=sgd.sgd(rnn,states,ground_truth)
    end=time.time()
    rnn.update({'W':0.5,'U':0.5,'V':0.5,'s':0.5})
    duration+=end-begin
    print i,'\t',err
print 'time: ', duration
